includes/class-wc-gateway-ppec-cart-handler.php

```diff
@@ -415,28 +415,44 @@ public function maybe_enqueue_checkout_js( $widget_title, $widget_instance = arr
 	 */
 	protected function get_button_settings( $settings, $context = '' ) {
 		$prefix = $context ? $context . '_' : $context;
-		$data = array(
+		$data   = array(
 			'button_layout'        => $settings->{ $prefix . 'button_layout' },
 			'button_size'          => $settings->{ $prefix . 'button_size' },
 			'button_label'         => $settings->{ $prefix . 'button_label' },
 			'hide_funding_methods' => $settings->{ $prefix . 'hide_funding_methods' },
 			'credit_enabled'       => $settings->{ $prefix . 'credit_enabled' },
 		);
 
+		if ( ! is_array( $data['hide_funding_methods'] ) ) {
+			$data['hide_funding_methods'] = array();
+		}
+
 		$button_layout        = $data['button_layout'];
 		$data['button_size']  = 'vertical' === $button_layout && 'small' === $data['button_size']
 			? 'medium'
 			: $data['button_size'];
 
 		if ( ! wc_gateway_ppec_is_credit_supported() ) {
 			$data['credit_enabled'] = 'no';
-			if ( ! is_array( $data['hide_funding_methods'] ) ) {
-				$data['hide_funding_methods'] = array( 'credit' );
-			} elseif ( ! in_array( 'credit', $data['hide_funding_methods'] ) ) {
+			if ( ! in_array( 'credit', $data['hide_funding_methods'] ) ) {
 				$data['hide_funding_methods'][] = 'credit';
 			}
 		}
 
+		// If we need to create a billing agreement or this is a subscription's single product page, hide credit/debit card options.
+		if (
+			! in_array( 'card', $data['hide_funding_methods'] )
+			&& ( 'single_product' !== $context && wc_gateway_ppec()->checkout->needs_billing_agreement_creation( array() ) )
```

Comment:
Checking that context is "single_product" is not enough to hide the buttons in all cases: PPEC only calls this function (`get_button_settings()`) with "single_product" as context when [you're on a single product page **and** the single product specific settings are enabled](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/3b40bb886be65102dcd09d8436436e1d971ce212/includes/class-wc-gateway-ppec-cart-handler.php#L514-L515).

If you're on a single product page but have no single product specific settings, this function is called with no context (default), but we still need to hide the buttons, which doesn't happen.

**Steps to test**:
1. Make sure "Configure settings specific to Single Product view" is not checked.
2. Visit a single product page.
3. Credit/debit card options are (incorrectly) shown.

- - -

Not sure if this means a filter implementation is more likely to work, as you can't rely on the context alone. Maybe [a filter around _this_ value](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/3b40bb886be65102dcd09d8436436e1d971ce212/includes/class-wc-gateway-ppec-cart-handler.php#L522) could work.

Reply:
> If you're on a single product page but have no single product specific settings, this function is called with no context (default), but we still need to hide the buttons, which doesn't happen.

I don't think the card fields are displayed in this case. 

For example these are my settings. 
<img width="584" alt="Screen Shot 2020-04-03 at 1 46 34 pm" src="https://user-images.githubusercontent.com/8490476/78322255-95234800-75b1-11ea-8915-180e2398ac89.png">

On a subscription page the credit card methods aren't displayed.

<img width="692" alt="Screen Shot 2020-04-03 at 1 47 32 pm" src="https://user-images.githubusercontent.com/8490476/78322318-b8e68e00-75b1-11ea-859b-a75c4febf258.png">

On a simple product they are.

<img width="639" alt="Screen Shot 2020-04-03 at 1 48 37 pm" src="https://user-images.githubusercontent.com/8490476/78322351-d74c8980-75b1-11ea-886e-1eee2a04813f.png">

This is handled by this condition. https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/hide_cards_for_subscriptions/includes/class-wc-gateway-ppec-cart-handler.php#L547-L553

Reply:
Hmm that's weird, that we're seeing different things. This works correctly for me when in SDK mode, but not in legacy mode. In fact, the condition you've mentioned is inside a `$use_js_sdk` so it doesn't apply to legacy mode.

I believe the problem is that in legacy mode we rely entirely on `get_button_settings()` and the "single_product" context, but this function doesn't get called with said context [unless there are single product specific settings](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/e0da0a4896e162c6866417aebd72868a1aeb9bc2/includes/class-wc-gateway-ppec-cart-handler.php#L514-L515) (which is separate from just enabling PPEC on this view).

I'm not sure why it works for you, though. Could you repeat the test again just to confirm?

Reply:
ok they have shown up this morning. 🤦‍♂